File: android/src/main/java/com/aptoslabs/arculus/csdk/CSDKAPI.kt
package com.aptoslabs.arculus.csdk

import android.nfc.tech.IsoDep
import com.sun.jna.Pointer

class CSDKAPI(private val tag: IsoDep) : AutoCloseable {
  private val wallet: Pointer = CSDK.WalletInit()

  override fun close() {
    CSDK.WalletFree(wallet)
  }

  suspend fun walletGetFirmwareVersion(): String {
    return WalletGetFirmwareVersion(wallet).execute(tag)
  }

  suspend fun walletGetGGUID(): String {
    return WalletGetGGUID(wallet).execute(tag)
  }

  suspend fun walletGetPublicKeyFromPath(path: String, curve: Short): Map<String, String> {
    return WalletGetPublicKeyFromPath(wallet, path, curve).execute(tag)
  }

  suspend fun walletInitRecoverWallet(nbrOfWords: Int) {
    return WalletInitRecoverWallet(wallet, nbrOfWords).execute(tag)
  }

  suspend fun walletInitSession() {
    return WalletInitSession(wallet).execute(tag)
  }

  suspend fun walletResetWallet() {
    return WalletResetWallet(wallet).execute(tag)
  }

  suspend fun walletSeedCreateWallet(nbrOfWords: Int): String {
    return WalletSeedCreateWallet(wallet, nbrOfWords).execute(tag)
  }

  suspend fun walletSeedFinishRecoverWallet(seed: ByteArray) {
    return WalletSeedFinishRecoverWallet(wallet, seed).execute(tag)
  }

  suspend fun walletSeedFromMnemonicSentence(mnemonicSentence: String): ByteArray {
    return WalletSeedFromMnemonicSentence(wallet, mnemonicSentence).execute()
  }

  private suspend fun walletSelectWalletV1() {
    val aid = WalletSelectWallet(
      wallet,
      WalletSelectWallet.ApplicationAID.V1
    ).execute(tag)

    if (!aid.contentEquals(WalletSelectWallet.ApplicationAID.V1.aid)) {
      throw ExpectedAIDNotMetException(WalletSelectWallet.ApplicationAID.V1)
    }
  }

  private suspend fun walletSelectWalletV2() {
    walletInitSession()
  }

  suspend fun walletSelectWallet() {
    val aid = WalletSelectWallet(
      wallet,
      WalletSelectWallet.ApplicationAID.V2
    ).execute(tag)

    when {
      aid.contentEquals(WalletSelectWallet.ApplicationAID.V1.aid) -> walletSelectWalletV1()
      aid.contentEquals(WalletSelectWallet.ApplicationAID.V2.aid) -> walletSelectWalletV2()
      else -> throw InvalidAIDException()
    }
  }

  suspend fun walletSignHash(path: String, curve: Short, algorithm: Byte, hash: String): String {
    return WalletSignHash(wallet, path, curve, algorithm, hash).execute(tag)
  }

  suspend fun walletStoreDataPIN(pin: String) {
    return WalletStoreDataPIN(wallet, pin).execute(tag)
  }

  suspend fun walletVerifyPIN(pin: String) {
    return WalletVerifyPIN(wallet, pin).execute(tag)
  }
}

File: ios/Arculus/CSDK/CSDKAPI.swift
import CoreNFC
import CSDK

class CSDKAPI {
    var tag: NFCISO7816Tag
    var wallet: OpaquePointer
    
    init(tag: NFCISO7816Tag) {
        self.tag = tag
        self.wallet = WalletInit()
    }
    
    deinit {
        WalletFree(wallet)
    }
    
    func walletGetFirmwareVersion() async throws -> WalletGetFirmwareVersion.ResponseType {
        try await WalletGetFirmwareVersion(wallet: wallet).execute(tag: tag)
    }
    
    func walletGetGGUID() async throws -> WalletGetGGUID.ResponseType {
        try await WalletGetGGUID(wallet: wallet).execute(tag: tag)
    }
    
    func walletGetPublicKeyFromPath(path: String, curve: UInt16) async throws -> WalletGetPublicKeyFromPath.ResponseType {
        try await WalletGetPublicKeyFromPath(wallet: wallet, path: path, curve: curve).execute(tag: tag)
    }
    
    func walletInitRecoverWallet(nbrOfWords: Int) async throws -> WalletInitRecoverWallet.ResponseType {
        try await WalletInitRecoverWallet(wallet: wallet, nbrOfWords: nbrOfWords).execute(tag: tag)
    }
    
    func walletInitSession() async throws -> WalletInitSession.ResponseType {
        try await WalletInitSession(wallet: wallet).execute(tag: tag)
    }
    
    func walletResetWallet() async throws -> WalletResetWallet.ResponseType {
        try await WalletResetWallet(wallet: wallet).execute(tag: tag)
    }
    
    func walletSeedCreateWallet(nbrOfWords: Int) async throws -> WalletSeedCreateWallet.ResponseType {
        try await WalletSeedCreateWallet(wallet: wallet, nbrOfWords: nbrOfWords).execute(tag: tag)
    }
    
    func walletSeedFinishRecoverWallet(seed: [UInt8]) async throws -> WalletSeedFinishRecoverWallet.ResponseType {
        try await WalletSeedFinishRecoverWallet(wallet: wallet, seed: seed).execute(tag: tag)
    }
    
    private func walletSelectWalletV1() async throws -> Void {
        let aid = try await WalletSelectWallet(wallet: wallet, applicationAID: ApplicationAID.v1).execute(tag: tag)
        
        guard aid == ApplicationAID.v1.aid else {
            throw WalletSelectWalletError.expectedAIDNotMet(ApplicationAID.v1)
        }
    }
    
    private func walletSelectWalletV2() async throws -> Void {
        try await walletInitSession()
    }
    
    func walletSelectWallet() async throws -> Void {
        let aid = try await WalletSelectWallet(wallet: wallet, applicationAID: ApplicationAID.v2).execute(tag: tag)
        
        switch aid {
        case ApplicationAID.v1.aid:
            try await walletSelectWalletV1()
        case ApplicationAID.v2.aid:
            try await walletSelectWalletV2()
        default:
            throw WalletSelectWalletError.invalidAID
        }
    }
    
    func walletSignHash(path: String, curve: UInt16, algorithm: UInt8, hash: String) async throws -> WalletSignHash.ResponseType {
        try await WalletSignHash(wallet: wallet, path: path, curve: curve, algorithm: algorithm, hash: hash).execute(tag: tag)
    }
    
    func walletStoreDataPIN(pin: String) async throws -> WalletStoreDataPIN.ResponseType {
        try await WalletStoreDataPIN(wallet: wallet, pin: pin).execute(tag: tag)
    }
    
    func walletVerifyPIN(pin: String) async throws -> WalletVerifyPIN.ResponseType {
        try await WalletVerifyPIN(wallet: wallet, pin: pin).execute(tag: tag)
    }
    
    func seedFromMnemonicSentence(mnemonicSentence: String) async throws -> WalletSeedFromMnemonicSentence.ResponseType {
        try await WalletSeedFromMnemonicSentence(wallet: wallet, mnemonicSentence: mnemonicSentence).execute()
    }
}

